### ISSUE TITLE
BLE Transmitter stopped advertising

### ISSUE THREAD

**Issue by schildgenb** [home-assistant/android#1855]
**Home Assistant Android version:**
beta-1594-e881a8a-full
**Android version:**
12
**Phone model:**
Pixel 3XL
**Home Assistant version:**
2021.10.6
**Last working Home Assistant release (if known):**
202110.6
**Description of problem:**
BLE Transmitter has stopped advertising. This stopped with 2021.10.0 version of Android. My wife's phone stopped as well. I looked through the logs and found this 

10-29 23:23:03.281 16400 16400 D BluetoothAdapter: isLeEnabled(): ON

10-29 23:23:03.288 16400 16400 E BeaconTransmitter: Advertisement start failed, code: 3

10-29 23:23:04.368 16400 16588 D

**Traceback (if applicable, to get the logs you may refer to: https://companion.home-assistant.io/docs/troubleshooting/faqs/#android-crash-logs):**

```

```

**Screenshot of problem:**

**Additional information:**

**Comment by dshokouhi** [comment]
are you using esphome? I think there is an important update. This bug is very similar to https://github.com/home-assistant/android/issues/1764 which has the same error message. The error code is actually saying the service has already started. There were some changes done to the output of the transmitter so you may need to adjust your settings on the receiving end.

**Comment by schildgenb** [comment]
I am using ESPresense for room tracking. I tried updating the settings
there but still nothing. I have even downloaded scanners and none of them
would pick up my wife's phone as they had before the last update.

**Comment by dshokouhi** [comment]
> I am using ESPresense for room tracking. I tried updating the settings there but still nothing. I have even downloaded scanners and none of them would pick up my wife's phone as they had before the last update.


Did you also make sure to adjust the sensor settings as the power output had changed?

@Alfiegerner FYI in case you didn't see this 🙂

**Comment by schildgenb** [comment]
I adjusted all of the settings to try different various things on both the
app and ESPresense. I tried the lastest beta as well and nothing shows up.
I download beacon simulator just to make sure it was not my phone but that
worked straight away. I downloaded scanner apps on my wife's phone but
nothing shows up besides the beacon simulator.

**Comment by mloebl** [comment]
I'm having the same issue, BLE/ESPresense works great with BeaconScope, but no longer working for me at all with the HA Android app starting this week. Using beta-1575-e1cc6c8-full

**Comment by kmalinich** [comment]
Same issue here. I'm on beta-1600-c6a5231-full

**Comment by Alfiegerner** [comment]
Hi all, I have a PR in that restores old behaviour as default and exposes setting to change measured power value.  Hopefully that should resolve this issue.  There should be a debug build available soon to try.

**Comment by dshokouhi** [comment]
Here is the debug APK from the above PR: https://github.com/home-assistant/android/actions/runs/1410125267

Can those who have this issue please test this APK? It will install side by side the production version so it wont interfere :)

**Comment by mloebl** [comment]
Thank you for the quick turn around @dshokouhi !  I installed the new version (it's icon is red, so easy to tell), disabled BeaconScope (which has been solid for the last 24 hours.) BeaconScope is set to -59 in , and 10hz reporting.  I set the same UUID and -59 power with this build, but not seeing anything in Esprense still.

**Comment by schildgenb** [comment]
Same here I installed the apk and still nothing showing up. I scanned again
and it was not picking anything up even when changing settings.

**Comment by schildgenb** [comment]
I looked at the logs and found this

BeaconTransmitter: Cannot start advertising due to exception

11-02 17:52:49.884 21097 21132 E BeaconTransmitter:
java.lang.SecurityException: Need android.permission.BLUETOOTH_ADVERTISE
permission for ***@***.***: GattService
startAdvertisingSet

11-02 17:52:49.884 21097 21132 E BeaconTransmitter:     at
android.os.Parcel.createExceptionOrNull(Parcel.java:2425)

**Comment by dshokouhi** [comment]
thank you for checking the logs :) this error seems new from before right?

**Comment by schildgenb** [comment]
Yes this is a different error then before.

**Comment by dshokouhi** [comment]
ok new debug APK to try: https://github.com/home-assistant/android/actions/runs/1424258401

🤞 this time I was actually able to reproduce the same error regarding permissions so I am hopeful but since I don't use these tools would be good for someone else to confirm :)

**Comment by mloebl** [comment]
Hi @dshokouhi Still getting an exception on starting here, running Android 12 on a Pixel 6 Pro:

```
11-05 07:30:56.493 32712 32712 D LocBroadcastReceiver: Begin evaluating if location update should be skipped
11-05 07:30:56.493 32712 32712 D LocBroadcastReceiver: Received location that is 81 milliseconds old, 1636111856412 compared to 1636111856493 with source fused
11-05 07:30:56.493 32712 32712 D LocBroadcastReceiver: Duplicate location received, not sending to HA
11-05 07:30:58.443 32712  1686 E FA      : Invalid google_app_id. Firebase Analytics disabled. See https://goo.gl/NAOOOI. provided id: mobilesdk_app_id
11-05 07:31:05.428 32712 32751 D SensorWorker: Updating all Sensors.
11-05 07:31:05.430 32712 32754 I WM-Processor: Moving WorkSpec (dfbc2082-363f-4480-b993-f56c1c8ae0d9) to the foreground
11-05 07:31:05.437 32712 32712 I WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND cmp=io.homeassistant.companion.android.debug/androidx.work.impl.foreground.SystemForegroundService (has extras) }
11-05 07:31:05.525 32712 32751 D SensorReceiver: Nothing to update
11-05 07:31:05.527 32712  2060 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=dfbc2082-363f-4480-b993-f56c1c8ae0d9, tags={ io.homeassistant.companion.android.sensors.SensorWorker } ]
11-05 07:31:05.532 32712 32712 I WM-SystemFgDispatcher: Stopping foreground service
11-05 07:31:10.562 32712 32751 D SensorWorker: Updating all Sensors.
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: Work [ id=119e07e2-631f-4dbc-a9fe-576869a8cdb1, tags={ io.homeassistant.companion.android.sensors.SensorWorker } ] was cancelled
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: java.util.concurrent.CancellationException: Task was cancelled.
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1184)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:514)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:311)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
11-05 07:31:10.576 32712  2060 I WM-WorkerWrapper: 	at java.lang.Thread.run(Thread.java:920)
11-05 07:31:10.636 32712 32712 I WM-SystemFgDispatcher: Stopping foreground service
11-05 07:31:10.642 32712 32753 D SensorWorker: Updating all Sensors.
11-05 07:31:10.644 32712  2054 I WM-Processor: Moving WorkSpec (1a8857b5-c7b0-4940-91d7-1376e9b56187) to the foreground
11-05 07:31:10.650 32712 32712 I WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND cmp=io.homeassistant.companion.android.debug/androidx.work.impl.foreground.SystemForegroundService (has extras) }
11-05 07:31:10.774 32712 32753 D SensorReceiver: Nothing to update
11-05 07:31:10.776 32712 32739 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=1a8857b5-c7b0-4940-91d7-1376e9b56187, tags={ io.homeassistant.companion.android.sensors.SensorWorker } ]
11-05 07:31:10.781 32712 32712 I WM-SystemFgDispatcher: Stopping foreground service
11-05 07:31:11.451 32712 32712 D BeaconParser: Parsing beacon layout: m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24
11-05 07:31:11.462 32712 32712 D BluetoothAdapter: isLeEnabled(): ON
11-05 07:31:11.463 32712 32712 D BluetoothAdapter: isLeEnabled(): ON
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: Cannot start advertising due to exception
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: java.lang.SecurityException: Need android.permission.BLUETOOTH_ADVERTISE permission for android.content.AttributionSource@b8334a23: GattService startAdvertisingSet
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Parcel.createException(Parcel.java:2409)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Parcel.readException(Parcel.java:2392)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Parcel.readException(Parcel.java:2334)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.bluetooth.IBluetoothGatt$Stub$Proxy.startAdvertisingSet(IBluetoothGatt.java:1989)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.bluetooth.le.BluetoothLeAdvertiser.startAdvertisingSet(BluetoothLeAdvertiser.java:460)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.bluetooth.le.BluetoothLeAdvertiser.startAdvertisingSet(BluetoothLeAdvertiser.java:329)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.bluetooth.le.BluetoothLeAdvertiser.startAdvertising(BluetoothLeAdvertiser.java:167)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.bluetooth.le.BluetoothLeAdvertiser.startAdvertising(BluetoothLeAdvertiser.java:101)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at org.altbeacon.beacon.BeaconTransmitter.startAdvertising(BeaconTransmitter.java:213)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at org.altbeacon.beacon.BeaconTransmitter.startAdvertising(BeaconTransmitter.java:159)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.bluetooth.ble.TransmitterManager.startTransmitting(TransmitterManager.kt:62)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.sensors.BluetoothSensorManager.updateBLESensor(BluetoothSensorManager.kt:213)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.sensors.BluetoothSensorManager.requestSensorUpdate(BluetoothSensorManager.kt:109)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.sensors.SensorDetailFragment.onCreatePreferences$lambda-2$lambda-1(SensorDetailFragment.kt:142)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.sensors.SensorDetailFragment.$r8$lambda$XMUSI_7loU7HvSZWwY4Z9mlnm0Y(Unknown Source:0)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at io.homeassistant.companion.android.sensors.SensorDetailFragment$$ExternalSyntheticLambda5.onPreferenceChange(Unknown Source:2)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.Preference.callChangeListener(Preference.java:1118)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.TwoStatePreference.onClick(TwoStatePreference.java:67)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.Preference.performClick(Preference.java:1182)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.Preference.performClick(Preference.java:1166)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.SwitchPreference.performClick(SwitchPreference.java:195)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at androidx.preference.Preference$1.onClick(Preference.java:181)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.view.View.performClick(View.java:7441)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.view.View.performClickInternal(View.java:7418)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.view.View.access$3700(View.java:835)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.view.View$PerformClick.run(View.java:28676)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Handler.handleCallback(Handler.java:938)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Handler.dispatchMessage(Handler.java:99)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Looper.loopOnce(Looper.java:201)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.os.Looper.loop(Looper.java:288)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at android.app.ActivityThread.main(ActivityThread.java:7839)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at java.lang.reflect.Method.invoke(Native Method)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
11-05 07:31:11.466 32712 32712 E BeaconTransmitter: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
11-05 07:31:11.507 32712 32753 D SensorWorker: Updating all Sensors.
11-05 07:31:11.509 32712  2060 I WM-Processor: Moving WorkSpec (95340650-eb5b-4c27-8286-d5f5a86edf76) to the foreground
11-05 07:31:11.515 32712 32712 I WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND cmp=io.homeassistant.companion.android.debug/androidx.work.impl.foreground.SystemForegroundService (has extras) }
11-05 07:31:11.583 32712 32753 D BluetoothAdapter: isLeEnabled(): ON
11-05 07:31:11.591 32712 32712 E BeaconTransmitter: Advertisement start failed, code: 3
11-05 07:31:11.596 32712 32753 D UrlRepository: localUrl is: false and usesInternalSsid is: false
```

Looks like this is the issue, `android.permission.BLUETOOTH_ADVERTISE permission` as missing, is there a way as a user I can access this permission as not seeing it in Android

**Comment by dshokouhi** [comment]
You need to go back to the sensor and disable/enable. It won't auto grant it.

**Comment by mloebl** [comment]
I thought I did but it didn't prompt, let me try again

**Comment by mloebl** [comment]
I'm not getting prompted, I get prompted for other permissions when doing enabling All Sensors like location and notifications. I get no permission prompt enabling/disable BLE nor any Bluetooth permissions prompts at all. Does the fact I am using Android 12 affect this?

**Comment by dshokouhi** [comment]
Can you check the app info page in settings? The Bluetooth pop-up should mention the advertise request. If it didn't pop-up I wonder if you need to clear data on the debug build and try again?

**Comment by mloebl** [comment]
I cleared the storage completely, went back to debug app so had to log back in and enable everything again for permissions.  Still no prompt for bluetooth, this is what it's app permissions look like:
![image](https://user-images.githubusercontent.com/309719/140524305-087f3814-cb0d-48c4-9bc0-eaedbcb46328.png)

**Comment by mloebl** [comment]
Ok, I don't think I'm getting that exception anymore now that Nearby Devices shows up, but still now BLE device showing up. I now have 2 phones I can test with, and running BeaconScope on the 2nd phone, I can see the COVID tracker on my Pixel 6 Pro, but nothing from the HA app. Not seeing anything in the logs either yet for errors

**Comment by dshokouhi** [comment]
What happens if you enable/disable the transmitter in settings for the sensor?

**Comment by mloebl** [comment]
Ok so I moved the debug version over to my old phone (Pixel 4 XL running Android 12) and removed the normal HA Android app to make debugging easier and not to cross streams. I'm getting the permission error again on the old phone.  `java.lang.SecurityException: Need android.permission.BLUETOOTH_ADVERTISE permission`  I've reset the cache to make sure as well.  This time instead of enabling All Sensors, I did just the Bluetooth related ones. One observation seems the first time I enable BLE, I get that exception, but not subsequent attempts however it doesn't seem to be working.

What's also interesting is that I noticed the app permissions on this phone had Nearby Devices, but not Fit. So I enabled Steps Sensor in addition to BLE, and it prompted for permissions to add Physical Activity, which I am most certainly not seeing when enabling the BLE sensor.

**Comment by dshokouhi** [comment]
Did you try disabling and enabling the transmitter in the BLE settings like I mentioned? I think there might be an order of operations issue there so wanted to check on that.  Because I do see that error but when I disable/enablethe transmitter from the setting I no longer see that error.

![image](https://user-images.githubusercontent.com/1634145/140532430-dec0edfc-d867-4840-b0c9-0ea284899e7f.png)

Also location tracking now request bluetooth permissions due to setting for high accuracy mode so its possible that you had already granted them but we may be hitting some odd issue which is why I asked you to try to disable/enable the transmitter from this setting.

**Comment by mloebl** [comment]
I am toggling this one:
![image](https://user-images.githubusercontent.com/309719/140533803-9d7e7db7-2268-480f-9b27-09dc480f2598.png)

I do see in the flow if I clicked the Enabled to send the sensor data to HA, and go back and look at the main sensors, I see it's not transmitting.I then go in and toggle the above, and it says it's transmitting but not being seen (and no permissions dialog for android.permission.BLUETOOTH_ADVERTISE is presented.)  I just cleared the app data and did this path again to confirm.

I also realized you were correct, I think on some tests I was clicking the Enable the Sensor vs the "Enable Transmitter" without realizing it. After I force stopped HA, started it back up and walked the path, toggling the "Enable Transmitter" does indeed through the permissions exception. I apologize for the confusion.

**Comment by schildgenb** [comment]
Yep having the same issue still. I have gone through various steps to see
if it will work but same error each time.

**Comment by dshokouhi** [comment]
Ok I see now so if you grant the app location permissions during onboarding that also includes the permission for this sensor which is why you don't see the pop-up appearing like you said.  The bluetooth permission was renamed to Nearby Devices in Android 12 as well.  Still not sure why that error is coming up when we are indeed requesting the permission. I wonder if its an issue with the library? 

With that said I saw a minor order of operations issue in the PR so there is a new APK, although I don't think it solves the issue :/

https://github.com/home-assistant/android/actions/runs/1426627645

**Comment by mloebl** [comment]
This looks like it may be working; showed up both as a beacon in BeaconScope AND in Espresense :) Thank you!

**Comment by dshokouhi** [comment]
yea the original condition was not correct :) so it was an order of operations issue on my end oops the error is now gone from the logs too right?

**Comment by mloebl** [comment]
Correct, I don't see any exceptions in the logs now for BLE :)  I tried cycling the Transmitter option as well.

**Comment by dshokouhi** [comment]
awesome, yea that was a silly mistake on my end 😂  thank you for testing 🙏

**Comment by mloebl** [comment]
No worries! Sorry about the confusion on the toggle, would have saved some back and forth :man_facepalming:  

You think this will make it in for this weekend's build? Curious to see if it will stop beaconing when Android decides to doze it (like Beaconscope randomly has happen), or if HA's Android app will handle it better.

**Comment by dshokouhi** [comment]
> You think this will make it in for this weekend's build?

depends on code review honestly



> Curious to see if it will stop beaconing when Android decides to doze it (like Beaconscope randomly has happen), or if HA's Android app will handle it better.

it shouldnt because its part of our SensorWorker service which runs on a 15 minute interval (fastest they offer) , the reason why you see "Updating sensors" notification is because thats a requirement for foreground services which prevents the app from getting killed by the system.

**Comment by mloebl** [comment]
Perfect, thank you for all your hard work on this!

**Comment by schildgenb** [comment]
I can also confirm this is working as well with that apk. Thanks for you
work as well!
